resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1464,6 +1464,14 @@ private object Client extends Logging {
     (mainUri ++ secondaryUris).toArray
   }
 
+  /**
+   * Returns a list of local, absolute URLs representing the user classpath.
+   *
+   * @param conf Spark configuration.
+   */
+  def getUserClasspathUrls(conf: SparkConf): Array[URL] =
+    getUserClasspath(conf).map(entry => new URL("file:" + new File(entry.getPath).getAbsolutePath))
```

Comment:
maybe I'm missing it but this seems wrong to me for the executors.  On yarn you have to put the current directory on jars specified that don't have an absolute path, because they are downloaded into distributed cache in ./

https://github.com/apache/spark/pull/32810/files#diff-de769ef2877d691f385e30a3a8a8a5d151e70e4e5d2acb9f6a03cf920ccda0d4L198

Reply:
Doing `new File(entry.getPath).getAbsolutePath)` will handle the relative paths by converting them to absolute. The values we get back from `getUserClasspath` are relative, like `./foo.jar`, and so the `getAbsolutePath` properly converts them to an absolute path. The snippet you linked uses the following:
```
Client.buildPath(Environment.PWD.$(), uri.getPath())
```
This is needed because it lives within `ExecutorRunnable`, which executes on the _driver_, but needs to resolve the working directory on the _executor_, so the environment variable is put into the path by the driver and later resolved in the executor process. In this case, we can just use standard Java IO/NIO mechanisms to make the path absolute, since the code is executing in the same process that will consume the path.

However this made me realize I had a bug regarding `local` paths which use the `GATEWAY_ROOT_PATH`/`REPLACEMENT_ROOT_PATH` configs. Will explain further in the main thread.